File: smells/unrealistic/mockers-without-borders/mockers-without-borders.js
// Subject under test
var App = {}

App.payMerchants = function (startDate, endDate) {
  var transactions = App.fetch(startDate, endDate)
  var purchaseOrders = App.createPurchaseOrders(App.groupByMerchant(transactions))
  App.submit(purchaseOrders)
}

// Test
var td = require('testdouble')
module.exports = {
  beforeEach: function () {
    td.replace(App, 'fetch')
    td.replace(App, 'submit')
  },
  paysMerchantsWithTotals: function () {
    var transactions = [
      {merchant: 'Nike', desc: 'Shoes', amount: 119.20},
      {merchant: 'Nike', desc: 'Waterproof spray', amount: 10.10},
      {merchant: 'Apple', desc: 'iPad', amount: 799.99},
      {merchant: 'Apple', desc: 'iPad Cover', amount: 59.99}
    ]
    var startDate = new Date(2015, 0, 1)
    var endDate = new Date(2015, 11, 31)
    td.when(App.fetch(startDate, endDate)).thenReturn(transactions)

    App.payMerchants(startDate, endDate)

    td.verify(App.submit([
      {merchant: 'Nike', total: 129.30},
      {merchant: 'Apple', total: 859.98}
    ]))
  },
  afterEach: function () {
    td.reset()
  }
}

// Fake production implementations to simplify example test of subject
var _ = require('lodash')

App.fetch = function (startDate, endDate) {
  // Imagine something that hits a data store here
}

App.groupByMerchant = function (transactions) {
  return _.groupBy(transactions, 'merchant')
}

App.createPurchaseOrders = function (transactionsByMerchant) {
  return _.map(transactionsByMerchant, function (transactions, merchant) {
    return {
      merchant: merchant,
      total: _.sumBy(transactions, 'amount')
    }
  })
}

App.submit = function (purchaseOrders) {
  // Imagine something that hits a payment processor here
}

File: smells/unrealistic/mockers-without-borders/mockers-without-borders.rb
# Subject under test
class App
  def pay_merchants(start_date, end_date)
    transactions = fetch(start_date, end_date)
    purchase_orders = create_purchase_orders(group_by_merchant(transactions))
    submit(purchase_orders)
  end
end

# Test
class MockersWithoutBorders < SmellTest
  def setup
    @subject = App.new
    super
  end
  def test_pays_merchants_with_totals
    transactions = [
      { merchant: 'Nike', desc: 'Shoes', amount: 119.20 },
      { merchant: 'Nike', desc: 'Waterproof spray', amount: 10.10 },
      { merchant: 'Apple', desc: 'iPad', amount: 799.99 },
      { merchant: 'Apple', desc: 'iPad Cover', amount: 59.99 }
    ]
    start_date = Date.civil(2015, 1, 1)
    end_date = Date.civil(2015, 12, 31)

    stub(@subject, :fetch, transactions, [start_date, end_date]) do
      verify(@subject, :submit, [
        { merchant: 'Nike', total: 129.30 },
        { merchant: 'Apple', total: 859.98 }
      ]) do
          @subject.pay_merchants(start_date, end_date)
      end
    end
  end
end

# Fake production implementations to simplify example test of subject
class App
  def fetch(start_date, end_date)
    # Imagine something that hits a data store here
  end

  def group_by_merchant(transactions)
    transactions.group_by { |t| t[:merchant] }
  end

  def create_purchase_orders(transactions_by_merchant)
    transactions_by_merchant.map { |(merchant, transactions)|
      {
        merchant: merchant,
        total: transactions.map {|h| h[:amount] }.reduce(:+)
      }
    }
  end

  def submit(purchase_orders)
    # Imagine something that hits a payment processor here
  end
end
